hibernate-core/src/main/java/org/hibernate/internal/CacheImpl.java

```diff
@@ -323,18 +338,46 @@ public void evictQueries() throws HibernateException {
 
 	@Override
 	public Region getSecondLevelCacheRegion(String regionName) {
-		return allCacheRegions.get( regionName );
+		// If there is an EntityRegion and CollectionRegion with the same name,
+		// then the EntityRegion should be returned.
+		Region region = entityRegionMap.get(regionName);
+		if ( region != null ) {
+			return region;
+		}
+		region = collectionRegionMap.get( regionName );
+		if ( region != null ) {
+			return region;
+		}
+		// Check for another type of Region of name regionName
+		return getNonEntityNonCollectionRegions().get( regionName );
```

Comment:
`getNonEntityNonCollectionRegions` creates a copy of all the collections, let's lookup the other maps...

Reply:
OK, I'll make the suggested change and I'll move the code from ```getNonEntityNonCollectionRegions``` into ```getAllSecondLevelCacheRegions```. 

I'll also document that to keep ```getSecondLevelCacheRegion``` and ```getAllSecondLevelCacheRegions``` consistent, the order of lookups in ```getSecondLevelCacheRegion``` needs to be the opposite of the order in which ```Map``` contents are added to ```getAllSecondLevelCacheRegions```